### ISSUE TITLE
erratic relatedDependencies behaviour

### ISSUE THREAD

**Issue by valentijnscholten** [dependency-check/DependencyCheck#2608]
Hi,

I am seeing erratic behaviour when scanning the same set of artifacts over and over again.
This is when there are multiple artifacts with the same (vulnerable) (bundles) dependencies.

Situation:

app_a.war contains handlebars-2.0.0.js
app_b.war contains handlebars-2.0.0.js

Both wars are components in the same application, let's say a customer facing war and a operator facing war both using a common backend rest api (left out for simplicity).

Scanning with dependency check delivers XML report along the following lines:

```
<dependency isVirtual="false">
	<fileName>app_a.war: handlebars-2.0.0.js</fileName>
	<filePath>/var/lib/jenkins/workspace/middleware_develop/target/app_a.war/apidocs/lib/handlebars-2.0.0.js</filePath>
	<md5>501c421a9bc201f50c76c8d28af0cb36</md5>
	<sha1>f25d39dc72774e392d55d98dd1d1285b1e213809</sha1>
	<sha256>7cb481a09730ac4f570ec37702f2fa70ce197bec81100565c6817eb13666a796</sha256>
	<relatedDependencies>
		<relatedDependency>
			<filePath>/var/lib/jenkins/workspace/middleware_develop/anotherfolder/app_b.war/apidocs/lib/handlebars-2.0.0.js</filePath>
			<sha256>7cb481a09730ac4f570ec37702f2fa70ce197bec81100565c6817eb13666a796</sha256>
			<sha1>f25d39dc72774e392d55d98dd1d1285b1e213809</sha1>
			<md5>501c421a9bc201f50c76c8d28af0cb36</md5>
		</relatedDependency>
	</relatedDependencies>
	
```

When using vulnerability aggregators or reporting tools, this hides app_b.war from the results most of the time as these
tools do not usually look at relatedDependencies.

Problem is 2 that sometimes dependency check flips the order around:

```
<dependency isVirtual="false">
	<fileName>app_b.war: handlebars-2.0.0.js</fileName>
	<filePath>/var/lib/jenkins/workspace/middleware_develop/target/anotherfolder/app_b.war/apidocs/lib/handlebars-2.0.0.js</filePath>
	<md5>501c421a9bc201f50c76c8d28af0cb36</md5>
	<sha1>f25d39dc72774e392d55d98dd1d1285b1e213809</sha1>
	<sha256>7cb481a09730ac4f570ec37702f2fa70ce197bec81100565c6817eb13666a796</sha256>
	<relatedDependencies>
		<relatedDependency>
			<filePath>/var/lib/jenkins/workspace/middleware_develop/app_a.war/apidocs/lib/handlebars-2.0.0.js</filePath>
			<sha256>7cb481a09730ac4f570ec37702f2fa70ce197bec81100565c6817eb13666a796</sha256>
			<sha1>f25d39dc72774e392d55d98dd1d1285b1e213809</sha1>
			<md5>501c421a9bc201f50c76c8d28af0cb36</md5>
		</relatedDependency>
	</relatedDependencies>
```

Is this expected / intended behaviour? I myself have (implicit) expectations:

1) Different scanned wars/artifacts should always be reported seperately.

If that's not reasonable / fair, then at least:

2) Any grouping / relating should be constant / predictable.

If that’s not possible, then at least:

3) Output the filename property also for the related dependencies / artifacts and output also the related dependencies as vulnerable in the html report / summary to avoid false negatives.

In the first example above it’s almost impossible to extra the “app_b.war” as filename for the related dependency.

Are these reasonable / doable expectations, or should we just exptect aggregators to look at the md5/sha1/sha256 and all relatedDependencies to "ungroup" / "unrelate" these items?
It's not only aggregators, also the html report looks different making it hard for the human eye to detect that we are in fact looking at identical reports.

Didn't know if I should mark it as bug or question or feature request.

Valentijn

**Comment by jeremylong** [comment]
Thanks for the report. I know why this is seeming inconsistent and we will work on improving the consistency.  

1. This is currently expected behavior that the same JS file in multiple WARs would be bundled. However, I agree that this probably should not be the case.  We can add some checking to eh bundling mechanism to correct this behavior.
2. We are apparently missing some a sorting/ordering mechanism as part of the bundling that is causing this behavior. While we plan on correcting item 1 - I can see scenarios where this still might occur so we need to review the logic and ensure we have a consistent ordering.
3. Adding the fileName node can be added in addition to the changes. While that could be parsed, it is easy enough to make the output a little more verbose.

**Comment by valentijnscholten** [comment]
Thanks for the quick response, to be sure some extra info:
1) It's also happening for .jar files inside multiple wars (not saying you stated otherwise, just mentioning it)
2) I think ordering can be fixed as long as no wars are added or removed, but it would be acceptable if in that case something would change a bit. To be honest, fixing the ordering would already solve my immediate problem of findings opening/closing in defect dojo.
3) Nice.

**Comment by valentijnscholten** [comment]
Parsing the full filepaths in the reports turns out not be straightforward, so I went ahead and created a PR to add the filename for related dependencies to the xml+json reports. (Item 3 from the list above).
Please let me know if this is something that you could see merged, as I will be using it in my setup.

**Comment by jeremylong** [comment]
This should be resolved with the pending 6.0.0 release.

**Comment by valentijnscholten** [comment]
@jeremylong only item 3 is fixed. Item 1 and 2 (too much bundling and ordering) are not fixed as far as I am aware.
